sdk/python/lib/pulumi/output.py

```diff
@@ -385,3 +387,179 @@ def __init__(self):
+                # that it is not set at all in the post-processed class.
+                delattr(cls, name)
+            else:
+                setattr(cls, name, prop.default)
+
+    # Mark this class with the signifier and save the properties.
+    setattr(cls, signifier_attr, True)
+    setattr(cls, _PULUMI_PROPERTIES, props)
+
+    return props
+
+
+def _create_py_property(cls: type, a_name: str, pulumi_name: str, translate):
+    """
+    Returns Python property getter that looks up the value in a dict.
+    """
+
+    # If the class itself is a subclass of dict, get the value from itself,
+    # otherwise, get the value from a private _values attribute.
+    if issubclass(cls, dict):
+        def getter(self):
+            # Grab dict's `get` method instead of calling `self.get` directly
+            # in case the type has a `get` property.
+            return getattr(dict, "get")(self, translate(self, pulumi_name))
+    else:
+        def getter(self):
+            return getattr(self, _VALUES).get(translate(self, pulumi_name))
+
+    getter.__name__ = a_name
+    return property(getter)
+
+
+def input_type(cls: Type[T]) -> Type[T]:
+    """
+    Returns the same class as was passed in, but marked as an input type.
+    """
+
+    # Get the input properties and mark the class as an input type.
+    props = _process_class(cls, _PULUMI_INPUT_TYPE)
+
+    # Add a _to_dict func, if needed.
+    if not hasattr(cls, _TO_DICT):
+        _translate_property = getattr(cls, _TRANSLATE_PROPERTY, None)
+        if callable(_translate_property):
+            def _to_dict(self) -> dict:
+                # Return a copy of `self.__dict__` with translated keys.
+                return {
+                    _translate_property(self, props[k].name if k in props else k): v
+                    for k, v in self.__dict__.items()
+                }
+        else:
+            def _to_dict(self) -> dict:
+                # Return a copy of `self.__dict__`.
+                return {
+                    props[k].name if k in props else k: v
+                    for k, v in self.__dict__.items()
+                }
+        setattr(cls, _TO_DICT, _to_dict)
+    return cls
+
+
+def output_type(cls: Type[T]) -> Type[T]:
+    """
+    Returns the same class as was passed in, but marked as an output type.
+
+    Python property getters are created for each Pulumi output property
+    defined in the class.
+
+    If the class is not a subclass of dict and doesn't have an __init__()
+    method, an __init__() method is added to the class that accepts a dict
+    representing the outputs.
+    """
+
+    # Get the output properties and mark the class as an output type.
+    props = _process_class(cls, _PULUMI_OUTPUT_TYPE)
+
+    # Add an __init__() method that takes a dict (representing outputs) as an arg
+    # if the class isn't a subclass of dict and doesn't have an __init__() method.
+    if not issubclass(cls, dict) and "__init__" not in cls.__dict__:
```

Comment:
Similarly - will things work if the class already had an `__init__`?  Should we error?  Perhaps it's more pythonic to not fail and just leave this so users can customize it if they can reverse engineer the required contract?

Reply:
Got it - if there's precedent in core Python libraries, then happy to align with those.